File: week1/plusOne.js
/**
 * @param {number[]} digits
 * @return {number[]}
 */
var plusOne = function(digits) {
    let exp = 1;//进位，现在假设之前有一个进位就是加一
    for(var i = digits.length-1;i>=0;i--){
        let sum = digits[i]+exp;
        digits[i] = sum%10;//加法结果
        exp = ~~(sum/10);//加法进位
    }
    if(exp === 1){digits.unshift(1)}//结束后还有进位，需要前面加上1
    return digits
};
console.log(plusOne([0]));
console.log(plusOne([9]));

File: week1/plusOne.c
#include <stdio.h>
#include <stdlib.h>
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* plusOne(int* digits, int digitsSize, int* returnSize){
    int exp = 1;
    int *result = (int*)malloc(sizeof(int)*digitsSize);
    int i = digitsSize-1;
    for(;i>=0;i--){
        int sum = digits[i]+exp;
        result[i] = sum%10;
        exp = ~~(sum/10);
    }
    if(exp==1){//需要额外加上进位
        int *expand = (int*)malloc(sizeof(int)*(digitsSize+1));
        expand[0] = 1;
        int j=0;
        for(;j<digitsSize;j++){
            expand[j+1] = result[j];
        }
        free(result);
        returnSize[0] = digitsSize+1;
        return expand;
    }
    returnSize[0] = digitsSize;
    return result;
}
void main(){
    int arr[4] = {4,3,2,1};
    int *result;
    plusOne(arr,sizeof(arr),result);
    getchar();
}